### ISSUE TITLE
Error when reused HTTP connection has been closed

### ISSUE THREAD

**Issue by tallakt** [edgurgel/httpoison#130]
I have a GenServer that polls a service on HTTP every half hour. It seems HTTPoison wil resuse the HTTP conenction for each of these, and from time to time this connection has been closed by either partner (probably the service).

This lead me to implementing the following code:

```
defmodule NOAA.HTTPClient do
  require Logger
  @kjevik "ENCN"
  @base "http://weather.noaa.gov/pub/data/observations/metar/stations/"

  def get(station_code \\ @kjevik) do
    # The connection may have been closed in the meantime, in that case allow
    # for one retry before failing
    fun = fn -> HTTPoison.get("#{@base}#{station_code}.TXT") end
    case fun.() do
      {:ok, result} ->
        result
      {:error, :closed} ->
        Logger.info "METAR HTTP socket closed, reopen and try again"
        {:ok, result} = fun.()
        result
    end
  end
end
```

It seems this code is rather more a HTTPoison concern, if HTTPoison as an implementation detail decides to reuse the connection. If the end user of HTTPoison needs to deal with this circumstance, I feel a minimum would be to add an option in the function signature/interface so that he could enable this as a built in feature, eg:

```
{:ok, result} = HTTPoison.get("#{@base}#{station_code}.TXT", reopen_stale_connection: true)
```

**Comment by edgurgel** [comment]
I totally agree with you and this hackney issue when solved should help us with this error case: https://github.com/benoitc/hackney/issues/161

**Comment by Sireeni** [comment]
I'm writing a web crawler in which the process in charge of the http requests dies with this error
```
18:10:54.852 [error] GenServer MCrawl.Producer terminating
** (HTTPoison.Error) :closed
    (httpoison) lib/httpoison.ex:66: HTTPoison.request!/5
[…]
```
@edgurgel Do you have any idea about a fix/workaround to avoid the crash? I'm using HTTPoison 0.9.0.

**Comment by sunboshan** [comment]
I'm having the same error when simply send a GET request to Google Play.

```
iex> HTTPoison.get!("https://play.google.com")
** (HTTPoison.Error) :closed
    (httpoison) lib/httpoison.ex:66: HTTPoison.request!/5
```

It was working in the morning, not sure if Google Play change something on their end, or it's a HTTPoison issue.

**Comment by iamd3vil** [comment]
I am also facing the same error from yesterday. GET request to `google.com`, `play.google.com` or `news.google.com` are failing.

```
iex> HTTPoison.get("https://google.com/")
{:error, %HTTPoison.Error{id: nil, reason: :closed}}
```

I am using HTTPoison 0.9.0  and Elixir 1.3.2, OTP 19

**Comment by dongmaster** [comment]
This happens to me too.
```
iex> HTTPoison.get("https://www.youtube.com")
{:error, %HTTPoison.Error{id: nil, reason: :closed}}
```

**Comment by freiguy1** [comment]
Also to me. I believe it has something to do with updating elixir/erlang.

**Comment by dongmaster** [comment]
Funny that you should mention that, this seemed to happen after I updated Elixir and Erlang to 1.3.1 and OTP 19 respectively.

**Comment by freiguy1** [comment]
So I had a bit of pressure to get this working this morning, since I wrote an app for our company which monitors our deployment environments and all of our environments were appearing down for 60-80 people.  I fixed it by downgrading erlang to OTP 18.2.  This is what I did (on ubuntu):

`sudo apt-get remove --auto-remove esl-erlang erlang elixir`
`sudo apt-get install esl-erlang=1:18:2`
`sudo apt-get install elixir`

I'd use these commands with caution and customize them to your own needs.  I should also note that these apt-get commands work for me because I've added the erlang solutions repo to the ubuntu package manager as seen [here](http://elixir-lang.org/install.html).

It's really disappointing running your package manager's update call can potentially break a phoenix app at runtime.  Now I'll need to be wary when upgrading in the future.

**Comment by dongmaster** [comment]
@freiguy1 That should be ``` sudo apt-get install esl-erlang=1:18.2 ```
Note the change from : to . at the end there.

**Comment by iamd3vil** [comment]
I can also confirm that downgrading to OTP 18 solved this for me, but any idea why this is happening with OTP 19.

**Comment by proProbe** [comment]
Hello! Im also having this problem, namely for youtube. Did anyone find a solution to this other than downgrading erlang?

**Comment by sschneider1207** [comment]
Adding the following to my `config.exs` fixes this for me:
```
config :ssl, protocol_version: :"tlsv1.2"
```
It seems that OTP 19 has a bug with the SSL module and default protocol version it tries to use.

**Comment by sunboshan** [comment]
@sschneider1207 Thanks for your solution! It works!

Also, I've been upgrade to Elixir 1.3.2 and OTP 19 for a while and HTTPoison.get always working. Until I want to install some dependency using `mix deps.get`, it prompts asking me to run `mix local.hex`. After I did that HTTPoison stops working for https website.

**Comment by sschneider1207** [comment]
@sunboshan I originally was having this issue when I was using the `:ssl` module directly, I was pointed to a erlang-solutions JIRA ticket related to the SSL module and OTP 19 when I asked about it in the Erlang slack group.  I figured this was the same problem since hackney uses the ssl module and the error was pretty similar to what I was getting.  If you have an https URL that HTTPoison is not working with, you can verify by doing `ssl:connect('host.com', 443, []).` in an erlang shell, you should get `{error, closed}`

**Comment by edgurgel** [comment]
Thanks for sharing your findings! @sschneider1207

**Comment by DevL** [comment]
For future reference, I believe the JIRA ticket mentioned above is this one: https://bugs.erlang.org/browse/ERL-192

**Comment by cdeyoung** [comment]
@DevL  - According to the JIRA ticket you linked to, the issue in Erlang has been resolved.  However, HTTPoison still doesn't work on various sites using Erlang 19 and Elixir 1.3.3, even using @sschneider1207's suggested fix.  For example, when I create a new project with the following setting in mix.exs: 

` def application do
    [applications: [:logger, :httpoison]]
  end
`

` def application do
    [applications: [:logger, :httpoison]]
  end
`

and @sschneider1207's suggestion in config.exs

`config :ssl, protocol_version: :"tlsv1.2"
`

I still end up with:

`iex(1)> HTTPoison.start
{:ok, []}
iex(2)> HTTPoison.post!("https://test.search.windows.net", "")
** (HTTPoison.Error) :closed
    (httpoison) lib/httpoison.ex:66: HTTPoison.request!/5`

The only way I've been able to achieve any kind of consistent stability has been to downgrade back to Erlang 18.3.  If there is a way to work around this issue, I'd really appreciate knowing what it is.

Thanks for all the information.

**Comment by DevL** [comment]
@cdeyoung I've noticed that locking `hackney` at `1.6.1` is a workaround for some sites with poor SSL configuration as well. Sadly this counteracts the security fixes made in Hackney in `1.6.2` and onwards. :-(

**Comment by olieidel** [comment]
Interestingly, neither of the workarounds work for the url
`http://docs.microsofttranslator.com/oauth-token.html` ([API Documentation](http://docs.microsofttranslator.com/oauth-token.html)).

I locked `hackney` at `1.6.1` and set the config 
`config :ssl, protocol_version: :"tlsv1.2"`

Further, trying HTTPotion yields a similar error. This leaves us in the slightly awkward position of not being able to make requests to some sites in Elixir at all..

Possibly related: [hackney #362](https://github.com/benoitc/hackney/issues/362)

**Comment by edgurgel** [comment]
Closing this issue. We have now an updated README with information about the SSL configuration: https://github.com/edgurgel/httpoison/commit/42efbe463221df4fe57c8804adda5f8eb483ceb0

**Comment by ijunaid8989** [comment]
Anybody had any experience with HTTPoison POST closed errors instead of get? in version 0.10.0?

**Comment by suddenrushofsushi** [comment]
Hey @edgurgel just a quick note the Erlang "bug" (ERL-192) was closed August 2016 as "not a bug". Using the workaround is still a perfectly reasonable fix.

**Comment by edgurgel** [comment]
Cool thanks @suddenrushofsushi for the info!

**Comment by joshuataylor** [comment]
I'm seeing this happen, usually after the first POST, when using HTTP not HTTPS.

I've tried setting up a pool, but still the same issue, is there anything else I could try, or just wait until hackey is updated?

**Comment by hayesgm** [comment]
The original issue was about reopening stale connections and possibly adding an option `reopen_stale_connection`. Do we have any work-arounds for that issue? @tallakt

**Comment by tallakt** [comment]
Its been so long since I used this, I am not up to date anymore. I handled it downstream in my code, which is ok, if the behaviour was obvious and documented.

**Comment by hayesgm** [comment]
Thanks, I followed a similar path and just added retry logic which seems to resolve the issue. It would be good to know if this is a real issue with Hackney or an issue with the server I'm connecting to.

**Comment by regisbsb** [comment]
@hayesgm may I ask you what retry logic please?

**Comment by edgurgel** [comment]
AFAIK Connections being closed with HTTP 1.1 are pretty much impossible to solve.

That's why HTTP/2 has GOAWAY frame to announce that a connection will be closed (because of an error or not). This allows the client to know which requests must be retried etc.

**Comment by ldd** [comment]
I hit a similar error on `esl-erlang` 23 ( just released today, I think)

Fix by @freiguy1 worked like a charm

with one tiny change:
`sudo apt-get install -y esl-erlang=1:22.3.4-1`
